File: lib/c_renderer/sprite3d-bridge.c
#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include "dragonruby.h"
#include "sprite3d-bridge.h"
#include "matrix.h"
#include "vertex.h"
#include "sprite.h"
#include "body.h"
#include "scene.h"
#include "camera.h"
#include "renderer.h"

/* ---=== VERTEX : ====--- */
DRB_FFI
Vertex* bridge_new_vertex(float x,float y, float z) {
  return new_vertex(x, y, z);
}

DRB_FFI
void bridge_compute_world_coordinates(Vertex* v,Body* b) {
  compute_world_coordinates(v, b->world);
}

DRB_FFI
void bridge_compute_view_coordinates(Vertex* v,SCamera* c) {
  matrix_vector_mul(c->view_matrix, v->world, v->view);
  v->int_view_z = (uint16_t)v->view[2];
  v->processed = VIEW_COORDS_CALCULATED;
}

DRB_FFI
void bridge_vertex_reset(Vertex *v) {
  vertex_reset(v);
}





/* ---=== SPRITE : ====--- */
DRB_FFI
Sprite* bridge_new_sprite(float x, float y, float z,int width,int height,int scale,int file_index,int atlas_x,int atlas_y,int atlas_w,int atlas_h) {
  Vertex* vertex  = new_vertex(x, y, z);
  return new_sprite(vertex, (uint16_t)width, (uint16_t)height, (uint16_t)scale, (uint16_t)file_index, (uint16_t)atlas_x, (uint16_t)atlas_y, (uint16_t)atlas_w, (uint16_t)atlas_h);
}

DRB_FFI
void bridge_print_sprite(Sprite* s,int indent) {
  print_sprite(s, indent);
}

DRB_FFI
//void bridge_sprite_rotate(Sprite* s,float da) {
void bridge_sprite_rotate(Sprite* s,int da) {
  sprite_rotate(s, da);
}

DRB_FFI
//void bridge_sprite_rotate_absolute(Sprite* s,float a) {
void bridge_sprite_rotate_absolute(Sprite* s,int a) {
  sprite_rotate_absolute(s, a);
}

DRB_FFI
int* bridge_sprite_get_render_data(Sprite* s) {
  return s->render_data;
}





/* ---=== BODY : ===--- */
DRB_FFI
Body* bridge_new_empty_body(int sprite_count) {
  return new_empty_body((size_t)sprite_count);
}

DRB_FFI
Sprite** bridge_get_body_sprites(Body* b) {
  return get_body_sprites(b);
}

DRB_FFI
void bridge_print_body(Body* b,int indent) {
  print_body(b, indent);
}

DRB_FFI
void bridge_body_move_to(Body* b,float x,float y,float z) {
  body_move_to(b, x, y, z);
}

DRB_FFI
void bridge_body_translate(Body* b,float dx,float dy,float dz) {
  body_translate(b, dx, dy, dz);
}

DRB_FFI
void bridge_body_rotate_x(Body* b,float a) {
  body_rotate_x(b, a);
}

DRB_FFI
void bridge_body_rotate_y(Body* b,float a) {
  body_rotate_y(b, a);
}

DRB_FFI
void bridge_body_rotate_z(Body* b,float a) {
  body_rotate_z(b, a);
}

DRB_FFI
void bridge_body_reset_rotation(Body* b) {
  body_reset_rotation(b);
}





/* ---=== SCENE : ===--- */
DRB_FFI
Scene* bridge_new_scene(void) {
  return new_scene();
}

DRB_FFI
void bridge_scene_push_element(Scene* s,Body* b) {
  scene_push_element(s, b);
}

DRB_FFI
void bridge_print_scene(Scene* s,int indent) {
  print_scene(s, indent);
}

DRB_FFI
int bridge_scene_get_element_count(Scene* s) {
  return (int)scene_get_element_count(s);
}

DRB_FFI
int bridge_scene_get_sprite_count(Scene* s) {
  return (int)scene_get_sprite_count(s);
}





/* ---=== CAMERA : ===--- */
DRB_FFI
SCamera* bridge_new_camera(float x, float y, float z,float fx, float fy, float fz,float ux, float uy, float uz) {  
  float position[4]  = {  x,  y,  z, 1.0 };
  float forward[4]   = { fx, fy, fz, 1.0 };
  float up[4]        = { ux, uy, uz, 1.0 };

  return new_camera(position, forward, up);
}

DRB_FFI
void bridge_print_camera(SCamera *camera) {
  print_camera(camera);
}

DRB_FFI
void bridge_camera_move_to(SCamera* camera,float x, float y, float z) {
  float position[] = { x, y, z, 1.0 };
  camera_move_to(camera, position);
}

DRB_FFI
void bridge_camera_translate(SCamera* camera,float dx, float dy, float dz) {
  float offset[] = { dx, dy, dz, 1.0 };
  camera_translate(camera, offset);
}

DRB_FFI
void bridge_camera_rotate_x(SCamera* camera,float angle) {
  camera_rotate_x(camera, angle);
}

DRB_FFI
void bridge_camera_rotate_y(SCamera* camera,float angle) {
  camera_rotate_y(camera, angle);
}

DRB_FFI
void bridge_camera_rotate_z(SCamera* camera,float angle) {
  camera_rotate_z(camera, angle);
}

DRB_FFI
void bridge_camera_move_forward(SCamera* camera,float speed) {
  camera_move_forward(camera, speed);
}

DRB_FFI
void bridge_camera_move_right(SCamera* camera,float speed) {
  camera_move_right(camera, speed);
}

DRB_FFI
void bridge_camera_move_up(SCamera* camera,float speed) {
  camera_move_up(camera, speed);
}





/* ---=== RENDERER : ===--- */
DRB_FFI
Renderer* bridge_new_renderer(int width,int height,float near,float far,int max_sprites) {
  return new_renderer((uint16_t)width, (uint16_t)height, near, far, (size_t)max_sprites);
}

DRB_FFI
void bridge_print_renderer(Renderer* renderer) {
  print_renderer(renderer);
}

DRB_FFI
void bridge_render_scene(Renderer* renderer,SCamera* camera,Scene* scene) {
  render_scene(renderer, camera, scene);
}

DRB_FFI
Sprite** bridge_renderer_get_sorted_sprites(Renderer* renderer) {
  return renderer->sorted_sprites;
}

DRB_FFI
int bridge_renderer_get_sorted_sprites_count(Renderer* renderer) {
  return (int)sorted_sprites_count(renderer);
}

File: lib/sprite3d_bridge.rb
$gtk.ffi_misc.gtk_dlopen("sprite3d")

module Engine3D
  ### VERTICES :
  class Vertex
    attr_reader :c_vertex

    def initialize(x,y,z)
      @c_vertex = FFI::Sprite3D_C::bridge_new_vertex(x, y, z)
    end

    def vertex_reset
      FFI::Sprite3D_C::bridge_vertex_reset(@c_vertex)
    end

    def set_local_coordinates(x,y,z)
      FFI::Sprite3D_C::bridge_vertex_set_local_coordinates(@c_vertex, x, y, z)

    def compute_world_coordinates(body);
      FFI::Sprite3D_C::bridge_vertex_compute_world_coordinates(@c_vertex, body)
    end

    def compute_view_coordinates(camera)
      FFI::Sprite3D_C::bridge_vertex_compute_view_coordinates(@c_vertex, camera)
    end

    def get_local_coordinates
      FFI::Sprite3D_C::bridge_vertex_get_local_coordinates(@c_vertex) 
    end

    def get_world_coordinates
      FFI::Sprite3D_C::bridge_vertex_get_world_coordinates(@c_vertex) 
    end

    def get_view_coordinates
      FFI::Sprite3D_C::bridge_vertex_get_view_coordinates(@c_vertex) 
    end
  end


  ### SPRITES :
  def self.new_sprite(x,y,z,width,height,scale,file,atlas_x,atlas_y,atlas_w,atlas_h)
    FFI::Sprite3D_C::bridge_new_sprite(x,y,z,width,height,scale,file,atlas_x,atlas_y,atlas_w,atlas_h)
  end
  
  def self.print_sprite(s,indent)
    FFI::Sprite3D_C::bridge_print_sprite(s,indent)
  end


  ### BODIES :
  class Body
    attr_reader :c_body

    def initialize(sprites)
      @c_body   = FFI::Sprite3D_C::bridge_new_empty_body(sprites.length)

      c_sprites = FFI::Sprite3D_C::bridge_get_body_sprites(@c_body)
      sprites.each.with_index { |s,i| c_sprites[i] = s }
    end
    
    def body_sprites
      FFI::Sprite3D_C::bridge_get_body_sprites(@c_body)
    end
    
    def print(indent)
      FFI::Sprite3D_C::bridge_print_body(@c_body,indent)
    end
    
    def move_to(x,y,z)
      FFI::Sprite3D_C::bridge_body_move_to(@c_body, x, y, z)
    end
    
    def translate(dx,dy,dz)
      FFI::Sprite3D_C::bridge_body_translate(@c_body, dx, dy, dz)
    end
    # untested
    
    def rotate_x(angle)
      FFI::Sprite3D_C::bridge_body_rotate_x(@c_body, angle)
    end
    
    def rotate_y(angle)
      FFI::Sprite3D_C::bridge_body_rotate_y(@c_body, angle)
    end
    
    def rotate_z(angle)
      FFI::Sprite3D_C::bridge_body_rotate_z(@c_body, angle)
    end
    
    def reset_rotation
      FFI::Sprite3D_C::bridge_body_reset_rotation(@c_body)
    end
    # untested
  end



  ### SCENE :
  class Scene
    attr_reader :c_scene

    def initialize()
      @c_scene  = FFI::Sprite3D_C::bridge_new_scene()
    end

    def <<(element)
      FFI::Sprite3D_C::bridge_scene_push_element(@c_scene,element.c_body)
    end

    def print(indent)
      FFI::Sprite3D_C::bridge_print_scene(@c_scene,indent)
    end

    def element_count
      FFI::Sprite3D_C::bridge_scene_get_element_count(@c_scene)
    end

    def sprite_count
      FFI::Sprite3D_C::bridge_scene_get_sprite_count(@c_scene)
    end
  end



  ### CAMERA :
  class Camera
    attr_reader :c_camera

    def initialize(x,y,z,fx,fy,fz,ux,uy,uz)
      @c_camera = FFI::Sprite3D_C::bridge_new_camera(x, y, z, fx, fy, fz, ux, uy, uz)
    end

    def print
      FFI::Sprite3D_C::bridge_print_camera(@c_camera)
    end

    def move_to(x,y,z)
      FFI::Sprite3D_C::bridge_camera_move_to(@c_camera, x, y, z)
    end
    # unstested

    def translate(dx,dy,dz)
      FFI::Sprite3D_C::bridge_camera_translate(@c_camera, dx, dy, dz)
    end
    # unstested

    def rotate_x(angle)
      FFI::Sprite3D_C::bridge_camera_rotate_x(@c_camera, angle)
    end

    def rotate_y(angle)
      FFI::Sprite3D_C::bridge_camera_rotate_y(@c_camera, angle)
    end
    
    def rotate_z(angle)
      FFI::Sprite3D_C::bridge_camera_rotate_z(@c_camera, angle)
    end
    
    def move_forward(speed)
      FFI::Sprite3D_C::bridge_camera_move_forward(@c_camera, speed)
    end
    
    def move_right(speed)
      FFI::Sprite3D_C::bridge_camera_move_right(@c_camera, speed)
    end
    
    def move_up(speed)
      FFI::Sprite3D_C::bridge_camera_move_up(@c_camera, speed)
    end
    # unstested
  end



  ### RENDERER :
  class Renderer
    attr_reader :c_renderer,
                :atlases

    #def initialize(args,width,height,near,far,max_sprites)
    def initialize(args,width,height,near,far,max_sprites,atlases)
      @c_renderer = FFI::Sprite3D_C::bridge_new_renderer(width, height, near, far, max_sprites)

      @atlases    = atlases

      max_sprites.times do |i|
        args.outputs.static_sprites << Engine3D::RenderSprite.new(self, max_sprites - i - 1)
      end
    end
    
    def render_scene(camera,scene)
      FFI::Sprite3D_C::bridge_render_scene( @c_renderer,
                                            camera.c_camera,
                                            scene.c_scene )
    end
    
    def print
      FFI::Sprite3D_C::bridge_print_renderer(@c_renderer)
    end
    
    def sorted_sprites
      FFI::Sprite3D_C::bridge_renderer_get_sorted_sprites(@c_renderer)
    end
    
    def sorted_sprites_count
      FFI::Sprite3D_C::bridge_renderer_get_sorted_sprites_count(@c_renderer)
    end
  end

  class RenderSprite
    WIDTH             = 0
    HEIGHT            = 1
    SCALE             = 2
    DRAW_X            = 3
    DRAW_Y            = 4
    DRAW_WIDTH        = 5
    DRAW_HEIGHT       = 6
    ANGLE             = 7
    ATLAS_FILE_INDEX  = 8
    ATLAS_X           = 9
    ATLAS_Y           = 10
    ATLAS_W           = 11
    ATLAS_H           = 12

    attr_reader :index
    attr_sprite
  
    def initialize(renderer,index)
      @atlases    = renderer.atlases
      @c_sprites  = FFI::Sprite3D_C::bridge_renderer_get_sorted_sprites(renderer.c_renderer)
      @index      = index
    end
  
    def sprite
      self
    end
  
    def draw_override(ffi_draw)
      render_data = FFI::Sprite3D_C::bridge_sprite_get_render_data(@c_sprites[@index])
      ffi_draw.draw_sprite_3( render_data[DRAW_X],
                              render_data[DRAW_Y],
                              render_data[DRAW_WIDTH],
                              render_data[DRAW_HEIGHT],
                              @atlases[render_data[ATLAS_FILE_INDEX]],
                              render_data[ANGLE],
                              255, 255, 255, 255,
                              false, false,
                              nil, nil, nil, nil,
                              0.0, 0.0,
                              render_data[ATLAS_X],
                              render_data[ATLAS_Y],
                              render_data[ATLAS_W],
                              render_data[ATLAS_H] )
    end
  end
end
